src/plugins/matrixops/pad.cpp

```diff
@@ -0,0 +1,416 @@
+            if (is_list_operand_strict(*it))
+                HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                    "pad::pad_get_array",
+                    generate_error_message("syntax error"));
+            m(i, 0) = extract_scalar_integer_value_strict(std::move(*it++));
+            if (is_list_operand_strict(*it))
+                HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                    "pad::pad_get_array",
+                    generate_error_message("syntax error"));
+            m(i, 1) = extract_scalar_integer_value_strict(std::move(*it));
+            i++;
+        }
+        return primitive_argument_type{std::move(m)};
+    }
+    ///////////////////////////////////////////////////////////////////////////
+    hpx::future<primitive_argument_type> pad::eval(
+        primitive_arguments_type const& operands,
+        primitive_arguments_type const& args, eval_context ctx) const
+    {
+        if (operands.empty() || operands.size() < 3 || operands.size() > 4)
+        {
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "pad::eval",
+                generate_error_message(
+                    "the pad primitive requires three or four operands"));
+        }
+
+        for (std::size_t i = 0; i != operands.size(); ++i)
+        {
+            if (!valid(operands[i]))
+            {
+                HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                    "pad::eval",
+                    generate_error_message(
+                        "the pad primitive requires that all the "
+                        "arguments to be valid"));
+            }
+        }
+
+        auto this_ = this->shared_from_this();
+        return hpx::dataflow(hpx::launch::sync,
+            hpx::util::unwrapping([this_ = std::move(this_)](
+                                      primitive_arguments_type&& args)
+                                      -> primitive_argument_type {
+                if (!is_string_operand(args[2]))
+                    HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                        "pad::eval",
+                        this_->generate_error_message(
+                            "the padding mode should be a string"));
+                if (extract_string_value(args[2]) != "constant")
+                    HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                        "pad::eval",
+                        this_->generate_error_message(
+                            "the pad primitive has only been implemented "
+                            "for constant mode"));
+                if (args.size() > 3 && !is_numeric_operand(args[3]))
+                    HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                        "pad::eval",
+                        this_->generate_error_message(
+                            "the current implementation of the pad primitive "
+                            "requires the "
+                            "constant_values argument to be a scalar"));
```

Comment:
The error message asks for a scalar in `args[3]`, but the code just checks for any numeric array value.